Add support for PSBLAS distributed vectors

fdrmrc opened the pull request:
This PR adds initial support for PSBLAS distributed vector, using the workflow introduced in #18662 (see also https://github.com/dealii/dealii/pull/18662#issuecomment-3083294737). The interface follows the ones of existing linear algebra libraries (mainly PETSc).

The main difference lies in the vector assembly phase. Currently, the assembly semantics are implicitly determined during vector initialization via the C API, which occurs inside the `reinit()` functions. As a result, contributions are automatically summed during assembly. To maintain consistency with other linear algebra classes, an assertion is thrown if one attempts to call `compress(VectorOperation::insert)`.

A set of tests has been added. While there, I have updated the CMake configuration and MPI initialization/finalization routines.

FYI @luca-heltai @sfilippone @Cirdans-Home @pasquadambra

luca-heltai commented:
@bangerth, @kronbichler, let me know if you agree with this interface. Marco already has two more PRs in the pipeline, with SparsityPattern+Matrix support, and AMG4PSBLAS. 

We are currently discussing over the design for GPU computation, and it should not be too far away. For the moment, these are results with AMG4PSBLAS and psblas matrices in step-40:

![IMAGE 2025-10-17 14:28:49](https://github.com/user-attachments/assets/99f2d279-e0f0-4efd-ac10-9df874850508)

drwells commented on source/lac/psblas_vector.cc:
```diff
@@ -0,0 +1,681 @@
+
+
+#include "deal.II/base/exception_macros.h"
+#include "deal.II/base/exceptions.h"
```
```suggestion
#include <deal.II/base/exception_macros.h>
#include <deal.II/base/exceptions.h>
```

drwells commented on source/lac/psblas_vector.cc:
```diff
@@ -0,0 +1,681 @@
+      {
+        int err;
+        err = psb_c_dgefree(psblas_vector, psblas_descriptor.get());
+        Assert(err == 0, internal::ExcFreePSBLASVector(err));
```
```suggestion
        AssertNoThrow(err == 0, internal::ExcFreePSBLASVector(err));
```

drwells commented on source/lac/psblas_vector.cc:
```diff
@@ -0,0 +1,681 @@
+  {
+    if (state != internal::State::Default)
+      {
+        int err;
```
```suggestion
        int ierr;
```
we usually use `ierr` in the library for an integer error code.

drwells commented on source/lac/psblas_vector.cc:
```diff
@@ -0,0 +1,681 @@
+        // ... insert the ghost indices ...
+        const std::vector<types::global_dof_index> &ghost_indexes =
+          ghost_indices.get_index_vector();
+        const int number_of_ghost_indices = ghost_indexes.size();
```
```suggestion
        const auto number_of_ghost_indices = ghost_indexes.size();
```
here and elsewhere we shouldn't convert to narrower integer types (I think the CI will also catch this)

drwells commented on include/deal.II/lac/psblas_vector.h:
```diff
@@ -0,0 +1,1083 @@
+    DeclException1(ExcInitializePSBLASVector,
+                   int,
+                   << "An error with error number " << arg1
+                   << " occurred while initializing a PSBLAS vector.");
```
What was the motivation for making these exceptions internal?

drwells commented on include/deal.II/lac/psblas_vector.h:
```diff
@@ -0,0 +1,1083 @@
+#include <deal.II/base/index_set.h>
+#include <deal.II/base/types.h>
+
+#include "deal.II/lac/vector_operation.h"
```
```suggestion
#include <deal.II/lac/vector_operation.h>
```

Cirdans-Home commented:
> Nice work - I have some basic style comments.
> 
> What is the plan for GPUs with psblas? Will that work with Kokkos?

Together with @sfilippone and @pdambra we went a different way than Kokkos to get GPU support. All the relevant data-types on the Fortran side are created following a state-pattern design. Namely, you always do declaration of the type:
```Fortran
type(psb_dspmat_type)  :: a ! double matrix
type(psb_d_vect_type) :: x,b ! double vectors
type(psb_desc_type):: desc_a ! commuication data structure
```
Internally they may then have a data storage format that is decided (and possibly changed) at runtime; the first time when you assemble them-in the Deal.II parlance, when you call `compress` on them:
```Fortran
call psb_cdasb(desc_a,info,mold=imold) 
call psb_spasb(a,desc_a,info,mold=amold)
call psb_geasb(x,desc_a,info,mold=vmold)
call psb_geasb(b,desc_a,info,mold=vmold)
```
Where now those variables on which the inner data type is _molded_ are pointers to the inner data structure, e.g., you want to use an Ellpack format on GPU, you select:
```
type(psb_d_cuda_hlg_sparse_mat), target   :: ahlg
type(psb_d_vect_cuda), target         :: dvgpu
type(psb_i_vect_cuda), target         :: ivgpu
class(psb_d_base_sparse_mat), pointer :: amold => ahlg
class(psb_d_base_vect_type), pointer :: vmold => dvgpu
class(psb_i_base_vect_type), pointer :: imold => ivgpu
```
If you want to use CSR on GPU you just change the matrix `mold`
```
type(psb_d_cuda_csrg_sparse_mat), target   :: acsrg
class(psb_d_base_sparse_mat), pointer :: amold => acsrg
```
While the part for the vectors has only to decide between CPU or GPU. Then you have few routine called `mv_to_fmt` (move to format) or `cp_to_fmt` (copy to format), to play around.

This has to be done consistently also with the preconditioners by passing the right `mold=` option to the routine assembling the precondioner, and its inner _work vectors_.

At this stage all this GPU construction works with NVIDIA GPUs/CUDA, but there is on the way the implementation of an internal data type which uses OpenACC to be more portable.

In the actual PR, the `mold` option is not exposed, i.e., the default is selected which is CSR on CPU.

We are discussing offline with @fdrmrc and @luca-heltai how to make a proposal of integration of this into Deal.II, but suggestions on how to make it more natural for Deal.II users are more than welcome.

kronbichler commented on include/deal.II/base/init_finalize.h:
```diff
@@ -60,7 +64,11 @@ enum class InitializeLibrary
-  P4EST = 32
+  P4EST = 32,
+  /**
+   * Initialize/finalize P4EST and SC.
```
```suggestion
   * Initialize/finalize PSBLAS.
```

luca-heltai commented:
> Thanks @drwells, I should have addressed all of your comments.

Would you mind fixing https://github.com/dealii/dealii/actions/runs/18631543573/job/53116948182?pr=18938 ?

Thanks!

fdrmrc commented:
> > Thanks @drwells, I should have addressed all of your comments.
> 
> Would you mind fixing https://github.com/dealii/dealii/actions/runs/18631543573/job/53116948182?pr=18938 ?
> 

Fixed. The CI failure is related to CMake's modules feature

drwells reviewed (approved):
Looks good. We still need to fix the module problem.
